src/meta/processors/job/BalanceJobExecutor.cpp

```diff
@@ -29,5 +47,1184 @@ folly::Future<Status> BalanceJobExecutor::executeInternal(HostAddr&& address,
+    auto host = it->first;
+    auto zoneIter =
+        std::find_if(zoneHosts.begin(), zoneHosts.end(), [host](const auto& pair) -> bool {
+          return host == pair.first.first;
+        });
+
+    if (zoneIter == zoneHosts.end()) {
+      LOG(INFO) << it->first << " have lost";
+      continue;
+    }
+
+    auto& hosts = zoneIter->second;
+    auto name = zoneIter->first.second;
+    for (auto hostIter = hosts.begin(); hostIter != hosts.end(); hostIter++) {
+      auto partIter = hostParts.find(*hostIter);
+      if (partIter == hostParts.end()) {
+        zoneParts_[it->first] = ZoneNameAndParts(name, std::vector<PartitionID>());
+      } else {
+        zoneParts_[it->first] = ZoneNameAndParts(name, partIter->second);
+      }
+    }
+  }
+  return nebula::cpp2::ErrorCode::SUCCEEDED;
+}
+
+bool DataBalanceJobExecutor::checkZoneLegal(const HostAddr& source, const HostAddr& target) {
+  VLOG(3) << "Check " << source << " : " << target;
+  auto sourceIter = std::find_if(zoneParts_.begin(), zoneParts_.end(), [&source](const auto& pair) {
+    return source == pair.first;
+  });
+
+  if (sourceIter == zoneParts_.end()) {
+    LOG(INFO) << "Source " << source << " not found";
+    return false;
+  }
+
+  auto targetIter = std::find_if(zoneParts_.begin(), zoneParts_.end(), [&target](const auto& pair) {
+    return target == pair.first;
+  });
+
+  if (targetIter == zoneParts_.end()) {
+    LOG(INFO) << "Target " << target << " not found";
+    return false;
+  }
+
+  LOG(INFO) << sourceIter->second.first << " : " << targetIter->second.first;
+  return sourceIter->second.first == targetIter->second.first;
+}
+
+nebula::cpp2::ErrorCode DataBalanceJobExecutor::prepare() {
+  auto activeHostsRet = ActiveHostsMan::getActiveHosts(kvstore_);
+  if (!nebula::ok(activeHostsRet)) {
+    auto retCode = nebula::error(activeHostsRet);
+    LOG(ERROR) << "Get active hosts failed, error: " << apache::thrift::util::enumNameSafe(retCode);
+    return retCode;
+  }
+  auto hosts = std::move(nebula::value(activeHostsRet));
+
+  if (hosts.empty()) {
+    LOG(ERROR) << "There is no active hosts";
+    return nebula::cpp2::ErrorCode::E_NO_HOSTS;
+  }
+  lostHosts_.reserve(paras_.size() - 1);
+  for (size_t i = 0; i < paras_.size() - 1; i++) {
+    lostHosts_.emplace_back(HostAddr::fromString(paras_[i]));
+  }
+  return nebula::cpp2::ErrorCode::SUCCEEDED;
+}
+
+nebula::cpp2::ErrorCode DataBalanceJobExecutor::finish(bool ret) {
+  std::lock_guard<std::mutex> lg(lock_);
+  return finishInternal(ret);
+}
+
+nebula::cpp2::ErrorCode DataBalanceJobExecutor::finishInternal(bool ret) {
+  CHECK(!lock_.try_lock());
+  plan_.reset(nullptr);
+  running_ = false;
+  auto rc = onFinished_(ret);
+  if (rc != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    return rc;
+  }
+  return ret ? nebula::cpp2::ErrorCode::SUCCEEDED : nebula::cpp2::ErrorCode::E_BALANCER_FAILURE;
+}
+
+folly::Future<Status> DataBalanceJobExecutor::executeInternal(HostAddr&& address,
```

Comment:
This is the origin `Balancer::balance`? Any logic changed? What is the `address` parameter?

Reply:
address is not useful here, it's  the MetaJobExecutor's parameter

Reply:
Maybe you could add another class instead of directly inherit from `MetaJobExecutor's`?
For example `MetaJobExecutor/StorageJobExecutor` both inherit from `JobExecutor`?

Because we don't need the parameter `address` in `JobManager::runJobInternal`. This current implementation do works, but weird. Including `TargetHosts::NONE`.

Reply:
You could think about it, if the change is quite big, could open another PR.

Reply:
ok